grpc-java/grpc-servlet-websocket-jakarta/src/main/java/io/grpc/servlet/web/websocket/WebsocketStreamImpl.java

```diff
@@ -0,0 +1,283 @@
+            if (logger.isLoggable(Level.FINE)) {
+                logger.log(
+                        Level.FINE,
+                        "[{0}] writeTrailers: {1}, headersSent = {2}, status = {3}",
+                        new Object[] {logId, trailers, headersSent, status});
+            }
+
+            // Trailers are always sent as asci, colon-delimited pairs, with \r\n separating them. The
+            // trailer response must be prefixed with 0x80 (0r 0x81 if compressed), followed by the size in a 4 byte int
+
+            byte[][] serializedTrailers = InternalMetadata.serialize(trailers);
+            // Total up the size of the payload: 5 bytes for the prefix, and each trailer needs a colon+space delimiter,
+            // and to end with \r\n
+            int trailerLength = Arrays.stream(serializedTrailers).mapToInt(arr -> arr.length + 2).sum();
+            ByteBuffer prefix = ByteBuffer.allocate(5);
+            prefix.put((byte) 0x80);
+            prefix.putInt(trailerLength);
+            prefix.flip();
+            ByteBuffer message = ByteBuffer.allocate(trailerLength);
+            for (int i = 0; i < serializedTrailers.length; i += 2) {
+                message.put(serializedTrailers[i]);
+                message.put((byte) ':');
+                message.put((byte) ' ');
+                message.put(serializedTrailers[i + 1]);
+                message.put((byte) '\r');
+                message.put((byte) '\n');
+            }
+            message.flip();
+            try {
+                // send in two separate messages
+                websocketSession.getBasicRemote().sendBinary(prefix);
+                websocketSession.getBasicRemote().sendBinary(message);
+
+                websocketSession.close();
+            } catch (IOException e) {
+                throw Status.fromThrowable(e).asRuntimeException();
+            }
+        }
+
+        @Override
+        public void cancel(Status status) {
+            if (!websocketSession.isOpen() && Status.Code.DEADLINE_EXCEEDED == status.getCode()) {
+                return; // let the servlet timeout, the container will sent RST_STREAM automatically
+            }
+            transportState.runOnTransportThread(() -> transportState.transportReportStatus(status));
+            // There is no way to RST_STREAM with CANCEL code, so write trailers instead
+            close(Status.CANCELLED.withCause(status.asRuntimeException()), new Metadata());
+            CountDownLatch countDownLatch = new CountDownLatch(1);
+            transportState.runOnTransportThread(() -> {
+                try {
+                    websocketSession.close();
+                } catch (IOException ioException) {
+                    // already closing, ignore
+                }
+                countDownLatch.countDown();
+            });
+            try {
+                countDownLatch.await(5, TimeUnit.SECONDS);
```

Comment:
Do we need to wait? Should we error if we timeout? Should we give the caller more control?

Reply:
This I did write - but i copied it from ServletServerStream.java, which did effectively the same thing in Sink.cancel(Status).

Unless I misunderstand, this is set up using a SerializingExecutor wrapped around a DirectExecutor, which means that the `transportState.runOnTransportThread` call itself will block until first the lock is available and second the lambda is finished. The countdownlatch is just to make sure that if another thread pool is used, we still get a chance to have the call complete first.

Note that the netty impl, in NettyServerStream.Sink.cancel() only enqueues the call and schedules a flush to execute it, but doesn't appear to wait for it to finish.